### PR TITLE
feat(shared): Organization hooks with `setData` and `revalidate`

### PR DESCRIPTION
**Pull request by panteliselef**
## Description

**Issue:** 
So far our organization hooks did not officially support mutations. Developers did not have a way to alternate the cache or simply revalidate on demand.

**Background:** 
For `invitationList` (from useOrganization) we didn't encounter such issues, because we are forcing an update under the hood when `lastOrganizationMember` changes. In the case of deleted resource, the hook would refetch and the deleted items would disappear from the cache. In the new flows we introduce with <OrganizationList/> we need to update the cache in place and avoid refetching, in order to avoid layout shifts.

We want developers to explicitly call either `revalidate` or `setData` after a resource is updated and let them decide which method best suits their use-case.

This PR exposes the mutate function for both useSWR and useSWR infinite based on the parameters a developer provided.
- ✅ Wrapper for swr mutate, allows us to have a consistent API without relying on swr not to break our customers.
- ✅ Wrapper for swr mutate, allows us to move to other internal implementation with ease.
- ✅ Can alternate local cache with `setData`
- ✅ Revalidate on demand with `revalidate`. (for example after creating an organization or accepting an invite)
- ❌ `mutate` is what swr describes as [Bound Mutate](https://swr.vercel.app/docs/mutation#bound-mutate) and does not use useSWRMutation under the hood.

As part of the PR, we replaced all usages of unstable__mutate from our update useOrganization and useOrganizationList hooks within ClerkJS Components
- `setData` will not trigger a revalidation, it will simply replace the passed data with data already in cache
- `revalidate` does not accept any parameter and will perform a revalidation causing a request to fire. 
    -  For infinite pagination it will refetch all the cached pages.  
    
Note: I attempted to optimize the revalidation of all pages with infinite by traversing the cache that  is only affected by the updated resource. 
There is also the case where this approach would result in duplicate keys when rendering lists.
This seems to be quite complicated as swr keeps n+1 records for infinite. That means the if 2 pages are fetched, 3 cache are created. 2 keys directly match the pages data and the 3rd is a "parent" record with its own key and it contains copies of the data of the 2 pages. Manually updating any of them, means we need to update the other as well.

## Checklist

- [x] `npm test` runs as expected.
- [x] `npm run build` runs as expected.
- [ ] (If applicable) [JSDoc comments](https://jsdoc.app/about-getting-started.html) have been added or updated for any package exports
- [x] (If applicable) [Documentation](https://github.com/clerkinc/clerk-docs) has been updated
This will be merged after the PR goes to production https://github.com/clerkinc/clerk-docs/pull/310

## Type of change

- [ ] 🐛 Bug fix
- [x] 🌟 New feature
- [ ] 🔨 Breaking change
- [x] 📖 Refactoring / dependency upgrade / documentation
- [ ] other:

## Packages affected

- [x] `@clerk/clerk-js`
- [ ] `@clerk/clerk-react`
- [ ] `@clerk/nextjs`
- [ ] `@clerk/remix`
- [ ] `@clerk/types`
- [ ] `@clerk/themes`
- [ ] `@clerk/localizations`
- [ ] `@clerk/clerk-expo`
- [ ] `@clerk/backend`
- [ ] `@clerk/clerk-sdk-node`
- [x] `@clerk/shared`
- [ ] `@clerk/fastify`
- [ ] `@clerk/chrome-extension`
- [ ] `gatsby-plugin-clerk`
- [ ] `build/tooling/chore`

### CODE REVIEW

**Comment by panteliselef** [comment]
!snapshot

**Comment by panteliselef** [comment]
!snapshot

**Comment by panteliselef** [comment]
!snapshot

**Comment by clerk-cookie** [comment]
Hey @panteliselef - the snapshot version command generated the following package versions:
| Package | Version |
| --- | --- |
| @clerk/backend | 0.29.1-snapshot.3ef2f60 |
| @clerk/chrome-extension | 0.4.0-snapshot.3ef2f60 |
| @clerk/clerk-js | 4.58.0-snapshot.3ef2f60 |
| @clerk/clerk-expo | 0.19.2-snapshot.3ef2f60 |
| @clerk/fastify | 0.6.7-snapshot.3ef2f60 |
| gatsby-plugin-clerk | 4.4.8-snapshot.3ef2f60 |
| @clerk/localizations | 1.26.0-snapshot.3ef2f60 |
| @clerk/nextjs | 4.24.0-snapshot.3ef2f60 |
| @clerk/clerk-react | 4.25.0-snapshot.3ef2f60 |
| @clerk/remix | 2.10.3-snapshot.3ef2f60 |
| @clerk/clerk-sdk-node | 4.12.6-snapshot.3ef2f60 |
| @clerk/shared | 0.22.1-snapshot.3ef2f60 |
| @clerk/types | 3.52.0-snapshot.3ef2f60 |

Tip: use the snippet copy button below to quickly install the required packages.
```sh
# @clerk/backend
npm i @clerk/backend@0.29.1-snapshot.3ef2f60
```
```sh
# @clerk/chrome-extension
npm i @clerk/chrome-extension@0.4.0-snapshot.3ef2f60
```
```sh
# @clerk/clerk-js
npm i @clerk/clerk-js@4.58.0-snapshot.3ef2f60
```
```sh
# @clerk/clerk-expo
npm i @clerk/clerk-expo@0.19.2-snapshot.3ef2f60
```
```sh
# @clerk/fastify
npm i @clerk/fastify@0.6.7-snapshot.3ef2f60
```
```sh
# gatsby-plugin-clerk
npm i gatsby-plugin-clerk@4.4.8-snapshot.3ef2f60
```
```sh
# @clerk/localizations
npm i @clerk/localizations@1.26.0-snapshot.3ef2f60
```
```sh
# @clerk/nextjs
npm i @clerk/nextjs@4.24.0-snapshot.3ef2f60
```
```sh
# @clerk/clerk-react
npm i @clerk/clerk-react@4.25.0-snapshot.3ef2f60
```
```sh
# @clerk/remix
npm i @clerk/remix@2.10.3-snapshot.3ef2f60
```
```sh
# @clerk/clerk-sdk-node
npm i @clerk/clerk-sdk-node@4.12.6-snapshot.3ef2f60
```
```sh
# @clerk/shared
npm i @clerk/shared@0.22.1-snapshot.3ef2f60
```
```sh
# @clerk/types
npm i @clerk/types@3.52.0-snapshot.3ef2f60
```

**Comment by SokratisVidros** [comment]
@panteliselef can you please elaborate on this? So far we have been using SWR internally for our own convenience. Did customers ask for this? In Clerk the preferred way to do mutations is via the Clerk models that hold all the mutating methods.

**Comment by panteliselef** [comment]
> @panteliselef can you please elaborate on this? So far we have been using SWR internally for our own convenience. Did customers ask for this? In Clerk the preferred way to do mutations is via the Clerk models that hold all the mutating methods.

Yes, there were customer requests, I'll share more internally.
I see SWR mutate as a way to force a revalidation or manually populate the cache for the collection of data. 
This does not change the way someone will call the mutation methods that update a resource. On the contrary those 2 mechanism will work together, let me give an example:

**Comment by SokratisVidros** [comment]
I see. Let's take it offline. We should be able to support this use case using the Clerk hooks API surface.

**Review by SokratisVidros — CHANGES_REQUESTED** [review]

File: packages/clerk-js/src/ui/components/OrganizationSwitcher/utils.ts
Comment (SokratisVidros) [reply packages/clerk-js/src/ui/components/OrganizationSwitcher/utils.ts:54]: 🤦 Thanks for the extra context. This needs a comment, as the future reader will be templated to update the code with a filter.

File: packages/shared/src/hooks/usePagesOrInfinite.ts
Comment (SokratisVidros) [reply packages/shared/src/hooks/usePagesOrInfinite.ts:219]: Not a fan at all. Do we have a use case for this already are can we defer it?

File: packages/clerk-js/src/ui/components/OrganizationSwitcher/utils.ts
Comment (SokratisVidros) [reply packages/clerk-js/src/ui/components/OrganizationSwitcher/utils.ts:27]: Gentle reminder ☝️

File: packages/clerk-js/src/ui/components/OrganizationSwitcher/utils.ts
Comment (SokratisVidros) [reply packages/clerk-js/src/ui/components/OrganizationSwitcher/utils.ts:54]: Gentle reminder ☝️

**Review by SokratisVidros — APPROVED** [review]

**Comment by dimkl** [comment]
@panteliselef We have changed the process of the v4 hotfixing. Please rebase this PR with `main` as the base branch and after it's closed, use the `backport` script to open a hotfix from `main-v4`.
Also if something that it's not related a fixing an issue in v4 and it can wait until v5, merge it in main and don't backport it.

File: packages/clerk-js/src/ui/components/OrganizationSwitcher/utils.ts
Comment (panteliselef) [reply packages/clerk-js/src/ui/components/OrganizationSwitcher/utils.ts:27]: If a page is undefined within the swr cache, it must remain undefined

File: .changeset/blue-ghosts-float.md
Comment (panteliselef) [reply .changeset/blue-ghosts-float.md]: I just updated this one

File: packages/shared/src/react/types.ts
Comment (dimkl) [inline packages/shared/src/react/types.ts:25]: ❓ shouldn't this be ` CacheSetter<(ClerkPaginatedResponse<T>[] | undefined)>` ?

Comment (panteliselef) [reply packages/shared/src/react/types.ts:25]: @dimkl a "page" in swr can be either undefined or has a value of type ClerkPaginatedResponse<T>

**Pull request merged**
